activation/atxdb.go

```diff
@@ -101,67 +123,90 @@ func (db *ActivationDb) CalcActiveSetFromView(a *types.ActivationTx) (uint32, er
-	if err == nil && len(prevAtxIds) > 0 {
-		prevAtx, err := db.GetAtx(prevAtxIds[len(prevAtxIds)-1])
-		if err == nil {
-			if prevAtx.Sequence >= atx.Sequence {
-				return fmt.Errorf("found atx with same seq or greater")
-			}
+	if atx.PrevATXId != *types.EmptyAtxId {
+		prevATX, err := db.GetAtx(atx.PrevATXId)
+		if err != nil {
+			return fmt.Errorf("prevATX not found")
+		}
+		if !prevATX.Valid {
+			return fmt.Errorf("prevATX not valid")
+		}
+		if prevATX.Sequence+1 != atx.Sequence {
+			return fmt.Errorf("sequence number is not one more than prev sequence number")
 		}
 	} else {
-		if prevAtxIds == nil && atx.PrevATXId != types.EmptyAtx {
-			return fmt.Errorf("found atx with invalid prev atx %v", atx.PrevATXId)
+		if atx.Sequence != 0 {
+			return fmt.Errorf("no prevATX reported, but sequence number not zero")
+		}
+		prevAtxIds, err := db.GetNodeAtxIds(atx.NodeId)
+		if err == nil && len(prevAtxIds) > 0 {
+			var ids []string
+			for _, x := range prevAtxIds {
+				ids = append(ids, x.String()[2:7])
+			}
+			return fmt.Errorf("no prevATX reported, but other ATXs with same nodeID (%v) found, atxs: %v", atx.NodeId.Key[:5], ids)
 		}
 	}
-	if atx.PositioningAtx != types.EmptyAtx {
+
+	if atx.PositioningAtx != *types.EmptyAtxId {
 		posAtx, err := db.GetAtx(atx.PositioningAtx)
 		if err != nil {
 			return fmt.Errorf("positioning atx not found")
 		}
 		if !posAtx.Valid {
 			return fmt.Errorf("positioning atx is not valid")
 		}
-		if atx.LayerIdx-posAtx.LayerIdx > NIPST_LAYERTIME {
-			return fmt.Errorf("distance between pos atx invalid %v ", atx.LayerIdx-posAtx.LayerIdx)
+		if atx.PubLayerIdx-posAtx.PubLayerIdx > db.LayersPerEpoch {
+			return fmt.Errorf("distance between pos atx invalid %v ", atx.PubLayerIdx-posAtx.PubLayerIdx)
 		}
 	} else {
-		if atx.LayerIdx/db.LayersPerEpoch != 0 {
+		publicationEpoch := atx.PubLayerIdx.GetEpoch(uint16(db.LayersPerEpoch))
+		if !publicationEpoch.IsGenesis() {
 			return fmt.Errorf("no positioning atx found")
 		}
 	}
 
-	//todo: verify NIPST challange
-	if atx.VerifiedActiveSet <= uint32(db.ActiveIds(types.EpochId(uint64(atx.LayerIdx)/uint64(db.LayersPerEpoch)))) {
-		return fmt.Errorf("positioning atx conatins view with more active ids than seen %v %v", atx.VerifiedActiveSet, db.ActiveIds(types.EpochId(uint64(atx.LayerIdx)/uint64(db.LayersPerEpoch))))
+	if atx.ActiveSetSize != atx.VerifiedActiveSet {
+		return fmt.Errorf("atx conatins view with unequal active ids (%v) than seen (%v)", atx.ActiveSetSize, atx.VerifiedActiveSet)
+	}
+
+	hash, err := atx.NIPSTChallenge.Hash()
+	if err != nil {
+		return fmt.Errorf("cannot get NIPST Challenge hash: %v", err)
+	}
+	db.log.Info("NIPST challenge: %v, OK nipst %v", hash.ShortString(), atx.NIPSTChallenge.String())
+	//todo: add validation of nipst
+	if err = db.nipstValidator.Validate(atx.Nipst, *atx.Nipst.NipstChallenge); err != nil {
+		return fmt.Errorf("NIPST not valid: %v", err)
+	}
+	if !atx.Nipst.ValidateNipstChallenge(hash) {
```

Comment:
If this validation necessary, after `nipstValidator` will be implemented?

Reply:
see comment above